### DISCUSSION TITLE
Custom ROM and locked bootloader

### DISCUSSION THREAD

**Discussion by MichaelZhuravsky** [chenxiaolong/avbroot#217] [Q&A]
Hello!

Thx for you work!

You've shown a way to sign modified OTA ROM, flash it and lock bootloader via fastboot.

Is there a difference to do the same with custom ROM like MIUI from xiaomi.eu? Seems some Xiaomi are allowed to use `fastboot flash avb_custom_key`, so I hope the it's possible. 

The custom ROMs usually need to be installed via recovery. Could you please help and show a step-by-step guide to make the custom rom's twrp-install-zip self-signed to lock bootloader?  

If it's not possible maybe fastboot's versions of these ROMs are more appropriate for this?

**Comment by chenxiaolong** [comment]
Glad you like avbroot!

The fastboot versions of the ROM won't work. You'll need the ones you flash from recovery. To determine if it's compatible, check if the zip contains `payload.bin` and `META-INF/com/android/metadata.pb` (or `metadata`) inside. If those files exist, then it should hopefully just work.

> Seems some Xiaomi are allowed to use `fastboot flash avb_custom_key`, so I hope the it's possible.

I've read some reports from 2021 that Xiaomi updated the bootloader on some devices where you can flash a custom key and relock the bootloader, but it will bootloop (until you unlock again). Hopefully that's not the case for your device. There's no way to know for sure without trying it.

**Comment by MichaelZhuravsky** [comment]
Thanks for the answer!

Yes, the TWRP-version of the ROM has these files.

My phone was originally released in 2020, i.e., there is a chance that the bootloader is not updated (does the update depend on the hardware, not the firmware?)

You said that if the phone launch fails with the locked bootloader, it will be possible to unlock it again. I hope so. I'm afraid of getting a hard brick :-)

Well, I'll try to gather my strength, find time and check it out. Tnx again!

**Reply by chenxiaolong** [reply]
> My phone was originally released in 2020, i.e., there is a chance that the bootloader is not updated (does the update depend on the hardware, not the firmware?)

Usually the bootloader updates are shipped with the Android updates. I'm not sure if it's the same for all Xiaomi devices, but this is an example of one I found: https://gitlab.com/CalyxOS/calyxos/-/issues/88 (last known bootloader that supported `avb_custom_key` on that device was bootloader version 9, shipped with Android 8). Maybe the version number shows up when you reboot into fastboot mode?

> You said that if the phone launch fails with the locked bootloader, it will be possible to unlock it again. I hope so. I'm afraid of getting a hard brick :-)

Yeah, that's definitely a valid fear :slightly_smiling_face: On Pixel devices, it's definitely possible to unlock again, but I don't have experience with any other manufacturer's devices.
